MSC2702: Specifying semantics for Content-Disposition on media (#2702)
matrix-org/matrix-spec-proposals: proposals/2702-media-content-disposition.md

for a client to link to `/thumbnail` directly, but in the event they do we provide a safe default.

When `Content-Disposition` is `inline`, the `Content-Type` SHOULD be one of the following mimetypes:

* `text/css`
* `text/plain`
* `text/csv`
* `application/json`
* `application/ld+json`
* `image/jpeg`
* `image/gif`
* `image/png`
* `image/apng`
* `image/webp`
* `image/avif`
* `video/mp4`
* `video/webm`
* `video/ogg`
* `video/quicktime`
* `audio/mp4`
* `audio/webm`
* `audio/aac`
* `audio/mpeg`
* `audio/ogg`
* `audio/wave`
* `audio/wav`
* `audio/x-wav`
* `audio/x-pn-wav`
* `audio/flac`
* `audio/x-flac`

If the content to be returned does *not* match these types, it SHOULD be returned as `attachment` (or in the
case of `/thumbnail`, not returned at all).

`Content-Type` additionally becomes a required header on both `/download` and `/thumbnail`, as `Content-Disposition`
without `Content-Type` is effectively useless in HTTP. The `Content-Type` header is the `Content-Type` supplied by
the client during `/upload`.

For clarity, a server is *not* required to use `inline` on `/download`. Clients SHOULD assume that a server will

> always use `attachment` instead.

Author: I'm not sure I see the difference between the texts there, but the expectation of a `SHOULD` earlier on in this proposal means that a server isn't required to follow the text surrounding. Namely, a server is *not* required to use `inline` in any particular scenario. This line of the MSC is simply trying to call that out to warn clients of the behaviour.

As for what the point of all this is: `Content-Disposition` can largely be seen as a security feature by server implementations. Normally we wouldn't call out that `SHOULD` means a server doesn't have to implement it, but considering servers might make the choice to use `attachment` out of caution (as Synapse and Dendrite both did for a little while), clients should be aware that the user experience may differ significantly if the underlying server implementation is different from their testing. Element Web/Desktop, for example, already does this by calling it "download file" rather than "view file" - it might open in a new tab and stay there (`inline`), or it might cause a download in the browser (`attachment`).

> always use `attachment` instead.

Author: (I've put words to that effect in the MSC)

> always use `attachment` instead.

Reviewer: The difference is one of emphasis. "Clients SHOULD assume that a server will always use `attachment` instead" means "it's not worth you writing any code to handle 'inline' because you're never going to see it in practice." "Clients SHOULD NOT rely on servers returning `inline` rather than `attachment`" means "normally it will be `inline` but you better be ready to receive `attachment` too".

The new wording is clear, thanks.

> always use `attachment` instead.

Reviewer 2: I don't really see the point of this paragraph since a Matrix client is not a browser. Why should it care about content-disposition?

> always use `attachment` instead.

Author: It's a reminder to client authors that the browser will have certain behaviour which may be undesirable. For example, an IRC bridge posting links to media may cause downloads to happen - the bridge should anticipate that rather than assume everything will appear inline.

* `Content-Disposition` SHOULD include a server-generated `filename`. For example, `thumbnail.png`.

Note that in both endpoints `Content-Disposition` becomes required, though the legal set of parameters is
intentionally different. Note also that by restricting `/thumbnail` to `inline` it reduces the allowed set
of `Content-Type` options the server can use, as per below.

When `Content-Disposition` is `inline`, the `Content-Type` MUST be one of the following mimetypes:

* `text/css`
* `text/plain`
* `text/csv`
* `application/json`
* `application/ld+json`
* `image/jpeg`
* `image/gif`
* `image/png`
* `image/apng`
* `image/webp`
* `image/avif`
* `video/mp4`
* `video/webm`
* `video/ogg`
* `video/quicktime`
* `audio/mp4`
* `audio/webm`
* `audio/aac`
* `audio/mpeg`
* `audio/ogg`
* `audio/wave`
* `audio/wav`
* `audio/x-wav`
* `audio/x-pn-wav`
* `audio/flac`
* `audio/x-flac`

If the content to be returned does *not* match these types, it MUST be returned as `attachment` (or in the
case of `/thumbnail`, not returned at all).

Servers SHOULD NOT rely on the `Content-Type` supplied to `/upload` to determine `Content-Disposition`. Instead,

> the content should be "sniffed" to determine appropriate type.

Reviewer 2: This is a hard problem to solve but not even OWASP gives concrete suggestions on how to tackle this. I think we can stress a bit that it's for layered security/defence in depth:

```
Servers SHOULD NOT rely on the `Content-Type` supplied to `/upload` to determine `Content-Disposition`. Instead,
the content should be "sniffed" to determine appropriate type as an extra protection together with the mimetype
management discussed above.
```

Author: I've clarified this differently, namely by saying that servers shouldn't use `inline` if they aren't sniffing for content.

Reviewer 2: They shouldn't use `inline` unless they force one of the safe content types in the allow list above. It doesn't matter if they sniff or not, especially since polyglot files are a thing.

Author: I believe that's already covered by the MSC?

Reviewer 2: Yes, it's the paragraph before this one.

# MSC2702: `Content-Disposition` usage in the media repo

The current specification does not clarify how to treat `Content-Disposition` on responses to
[`/download`](https://spec.matrix.org/v1.8/client-server-api/#get_matrixmediav3downloadservernamemediaid)
and [`/thumbnail`](https://spec.matrix.org/v1.8/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)
requests. Some clients expect attachments to be `download` only (or don't care), while other applications

> like bridges rely on it sometimes being `inline` for user experience reasons.

Reviewer 3: i'm failing to understand why bridges would ever care about content-disposition, let alone rely on it to be inline?

Reviewer 3: oooooh, this is so that if you click on a link to the media repo from an IRC bridge, it shows the contents in browser rather than chucking a download into your downloads dir?  if so, it would be great to actually say this in the problem statement :D

Author: The client (or bridge) can't control the header, but the IRC bridge for example does rely on the server using an `inline` disposition so it shows up in the browser.

Will clarify.

For `/thumbnail`:

* `Content-Disposition` MUST be returned, and MUST be `inline` (see below).
* `Content-Disposition` SHOULD include a server-generated `filename`. For example, `thumbnail.png`.

Note that in both endpoints `Content-Disposition` becomes required, though the legal set of parameters is
intentionally different. Note also that by restricting `/thumbnail` to `inline` it reduces the allowed set
of `Content-Type` options the server can use, as per below.

It's not typical for a client to link to `/thumbnail` in a way which makes use of `Content-Disposition`, but
in the event that a client does (and for consistency within the spec), we ensure a safe default is used.

When `Content-Disposition` is `inline`, the `Content-Type` MUST be one of the following mimetypes:

* `text/css`
* `text/plain`
* `text/csv`
* `application/json`
* `application/ld+json`
* `image/jpeg`
* `image/gif`
* `image/png`
* `image/apng`
* `image/webp`
* `image/avif`
* `video/mp4`
* `video/webm`
* `video/ogg`
* `video/quicktime`
* `audio/mp4`
* `audio/webm`
* `audio/aac`
* `audio/mpeg`
* `audio/ogg`
* `audio/wave`
* `audio/wav`
* `audio/x-wav`
* `audio/x-pn-wav`
* `audio/flac`

> * `audio/x-flac`

Reviewer 4: I'm finding it very hard to make a yes/no decision on this MSC in general, this part in particular. Where is this list from? Why only these types? Why does it need to be specced rather than allowing servers to pick something they think is sensible? What bad thing would happen if different homeservers had different lists?

Reviewer 4: Oh okay, I found answers to some of these questions in the Security section, but I still don't understand why it needs to be mandated as part of the spec.

Reviewer 2: It's just a [set of suggestions](https://github.com/matrix-org/matrix-spec-proposals/pull/2702/files#diff-aa3233268cd33ed1ed7a85d8ce397ca3357af7ec7adb626e569918f77a88d060R9).

Reviewer 4: Almost all of this, in particular the list of types I referred to, uses the word MUST. What parts are you referring to that are suggestions?

Author: It's a SHOULD for `inline`, but if the server does use it then added requirements are imposed.

Reviewer 5: >  When `Content-Disposition` is `inline`, the `Content-Type` MUST ...

I am also reading a MUST, not a SHOULD?

Author: From earlier in the MSC:

> For `/download`:
> 
> * `Content-Disposition` MUST be returned, and MUST be one of `inline` or `attachment` (see below).

> For `/thumbnail`:
> 
> * `Content-Disposition` MUST be returned, and MUST be `inline` (see below).
> * [cut]
> 
> Note that in both endpoints `Content-Disposition` becomes required, though the legal set of parameters is
> intentionally different. Note also that by restricting `/thumbnail` to `inline` it reduces the allowed set
> of `Content-Type` options the server can use, as per below.

This does make `inline` used for thumbnails, but all servers so far generate `image/png` or `image/jpeg` thumbnails, so are already compliant. The concern appears to be with teh `/download` endpoint, which has an `or` in there (implying SHOULD). 

This bit of the MSC is saying that if a server chooses to use `inline` for `/download`, and for all `/thumbnail` requests, then the returned `Content-Type` must be legal under the set. If the server chooses to use `attachment` (or "not inline") for `/download`, these `Content-Type` restrictions do not apply to them.

Reviewer 4: I feel like we may have got lost down a rabbit hole chasing an extremely specific semantic point here. This doesn't feel very relevant to me: regardless of this specific one, there are still a whole bunch of MUSTs in here which surely makes it something other than a "a set of suggestions".

I still don't really understand why a server that has a need to return another content type and it's been established with complete confidence that it's safe to do so, why they cannot do so without behaving out of spec.

Author: I don't believe we trust that servers can make that judgement call on their own, considering they've got it wrong (after doing research). We might be able to relax this, and maybe this is a sign that the endpoint is overloaded. In previous drafts of this MSC it was discussed whether it's correct that a user lands on `/download` in their browser - we weren't able to really conclude that at the time, so the current version assumes a combined purpose endpoint.

Is there a case where a server *wants* to return something not on the list as inline?

Reviewer 6: > Is there a case where a server _wants_ to return something not on the list as inline?

There's always going to be other potential things that are "safe", e.g. other image formats (JPEG 2000 is `image/jp2` among others); or audio formats (MIDI is `audio/midi`). 🤷

Reviewer 4: Okay, so there's the point: for sure implementers will get it wrong in ways that impact security. However, I don't believe the way to save them from this is to provide a rigid set of rules that must be followed with no explanation in order to be spec compliant. I think we absolutely can and should give advice and recommendations to help implementors make good, informed decisions though. So for example in this case, perhaps we could put this list as a, "this list is selected on this basis and is known-safe: if you want to return other types inline, here's are the things you need to be sure about in order to keep your users safe".

Reviewer 6: More importantly shouldn't it be highlighted the ones which are known **unsafe**?

Reviewer: Just adding a +1 to @dbkr's/@clokep's comments. It's entirely reasonable for the spec to give advice on a recommended list (or a list of types to avoid). Claiming that we as protocol maintainers know better than all present and future server maintainers seems somewhat arrogant.

Author: > More importantly shouldn't it be highlighted the ones which are known **unsafe**?

Listing unsafe ones would potentially be more risky: if we forget one, the default assumption from a developer would be that it's *safe* rather than unsafe. By listing the safe mimetypes, the default assumption is that it's *unsafe* (hopefully, with wording changes).

---

Ack on wording changes needed. Is this a concern that should block FCP?

Reviewer: > Is this a concern that should block FCP?

I don't feel like I want to give it a tick at the moment.

Author: That should be listed as a concern in that case, imo.

* `audio/wav`
* `audio/x-wav`
* `audio/x-pn-wav`
* `audio/flac`

> * `audio/x-flac`

Reviewer 7: For the sake of anyone (like me) who's surprised that `audio/opus` isn't on the list, the `audio/opus` mimetype seems to be [specifically for RTP payloads](https://datatracker.ietf.org/doc/html/rfc7587#section-6.1).  Recorded files (such as what we would use in voice messages) would be served with the `audio/ogg` mimetype.

Reviewer 4: Yeah, flac has its own container format for some reason. Opus just lives in ogg/webp/mp4 (it's the hermit crab of codecs).

the content should be "sniffed" to determine appropriate type. To effect this layered protection, servers which

> do not sniff for `Content-Type` SHOULD NOT use `Content-Disposition: inline`.

Reviewer: Presumably this is to stop servers presenting things which are *obviously* not `text/css` as `text/css`, or something? I mean, it can't be much of a security defence because anyone who is determined to disguise a file as something else will probably manage it?

A word or two to explain this might be appreciated.

Author: ftr, `Content-Disposition` effectively implies `Content-Type` being required (otherwise `Content-Disposition` doesn't do anything). With `Content-Type` required, the server should be putting in *some* effort to ensure it's actually serving what it thinks it's serving.

It's true that it can be worked around, but good is better than perfect here.

Reviewer: Sure, I'm just after some clarification about what we're trying to achieve here.

Reviewer 4: To help me understand what's happened here: we're resolving this comment by changing the MSC (the the 'fix wording' commit) from advising sniffing to advising using the the user-supplied content-type?

Author: Yes - content type sniffing is nearly impossible to get right, and it turns out that it doesn't protect anything if the server does get it right anyways. The Content Security Policy will kick in as well as the browser's natural sandboxing on any media which presents itself as an improper type.

Reviewer 4: Indeed, thanks: my problem wasn't in understanding why sniffing is bad, it was in trying to understand why the MSC proposed what it proposed at various points and why/when it changed between the two commits both marked, "wording".

> do not sniff for `Content-Type` SHOULD NOT use `Content-Disposition: inline`.

Reviewer: also, am I right in believing this is going to represent a change in behaviour for Synapse (which does not inspect file contents to determine content-types, and I suspect it is unlikely to)? Is that going to cause breakage in practice?

Author: It won't break anything - it's just a regular bug in Synapse. This is a SHOULD for that reason.

Reviewer: I'm not entirely following. To rephrase my question: I understand that Synapse is free to ignore this part of the MSC, but assuming it does not ignore it (and also does not start scanning content), will it cause breakage?

Author: There's a lot of conditions in there, sorry. If Synapse decides to not scan media, and also decides to embrace the security charactertistics of this MSC (namely serving everything as `attachment`), there will not be breakage that this MSC is concerned about. Clients relying on unspecced behaviour of Synapse may be affected, but that would be an implementation-level decision in either/both the client and server to work out.

It is my personal suggestion that Synapse scans media, at least roughly, to identify common file types. For example, `image/png` and `text/plain`.

Reviewer 2: Detecting filetypes is a hard problem, that's why files should be served from a different origin.
If the server enforces a `Content-Type`, the client will honour that settings (except ancient Internet Explorer versions from 10-15 years ago) so it can serve a text/html as text/plain with no risks, modulo bugs in the client.

Reviewer: @davidegirardi Are you proposing any particular changes to the MSC here? I think I might be missing your point.

Reviewer 2: I'd avoid talking about content detection/sniffing and instead propose to stick to an allow-list of safe content-types.

Reviewer: > I'd avoid talking about content detection/sniffing and instead propose to stick to an allow-list of safe content-types.

Sorry, I'm still not really following. You're suggesting that servers should not scan content, and instead should trust clients to be honest about the content type at upload time?

It sounds orthogonal to my question so should probably be a separate thread?

Author: (it also sounds like a request to have media be on a different domain from the homeserver, which would be another MSC's territory)

Reviewer 2: > Sorry, I'm still not really following. You're suggesting that servers should not scan content, and instead should trust clients to be honest about the content type at upload time?

No, I am saying that scanning the content will only give a mid-low degree of confidence regarding the actual file type so the real line of defence here is choosing which mimetypes are OK for content disposition inline.

@turt2live yes, using another origin is for another MSC but it gives context to what I'm writing I think.

Reviewer: > No, I am saying that scanning the content will only give a mid-low degree of confidence regarding the actual file type so the real line of defence here is choosing which mimetypes are OK for content disposition inline.

Ok, so the point is that you think it is ok to serve a file with `Content-Disposition: inline` even if the server hasn't scanned the content, provided it has a known-safe mimetype?

Reviewer 2: Yes because the content type specified by the server will determine what the browser will do.

Reviewer: Returning to the original point of this thread: I was trying to get an insight from those familiar with the area as to whether implementing this MSC will cause breakage in practice.

My impression is that it will, and indeed it it seems like the Synapse team have already tried it and had to roll it back: https://github.com/matrix-org/synapse/pull/15988.

That isn't to say we shouldn't drive through the necessary changes if we think it valuable; I'm just trying to get recognition of what the necessary changes are.

Reviewer 6: > My impression is that it will, and indeed it it seems like the Synapse team have already tried it and had to roll it back: [matrix-org/synapse#15988](https://github.com/matrix-org/synapse/pull/15988).

I don't think this is accurate -- Synapse [made a change](https://github.com/matrix-org/synapse/pull/15680) to *always return* `attachment`; this [wasn't very popular with those using  bridges](https://github.com/matrix-org/synapse/issues/15885); we then [implemented similar guidelines](https://github.com/matrix-org/synapse/pull/15988) to this MSC, except:

* No content sniffing is done (as @davidegirardi suggests)
* No changes were made to the thumbnail endpoint (although I think by default it will match this MSC since Synapse will return an error if it can't create thumbnail, and can only thumbnail safe image formats).
* I haven't double checked that the lists are identical.

I'm not sure if that answers your questions though?

Reviewer: > Synapse made a change to always return attachment; this wasn't very popular with those using bridges;

This is my point. The net effect of this MSC is to recommend that Synapse return attachment (or that it start scanning files). In practice that doesn't work, so it's a bit of a spurious recommendation.

Reviewer 6: > The net effect of this MSC is to recommend that Synapse return attachment (or that it start scanning files)

But we *do* return `attachment`, but only for "unsafe" content type.

("Unsafe" being anything not on an allow list.)

Reviewer: > only for "unsafe" content type.

It is the fact we do not do it for "safe" content types that is against the recommendation of the MSC, AFAICT. The MSC recommends we return `attachment` for *all* content types.

Reviewer 6: I'm confused, the MSC says:

> `Content-Disposition` MUST be returned, and MUST be one of `inline` or `attachment` (see below)

> When `Content-Disposition` is `inline`, the `Content-Type` MUST be one of the following mimetypes:
>
> [...list of mimetypes...]

This doesn't seem to match that we need to return `attachment` always?

Reviewer: > To effect this layered protection, servers which do not sniff for `Content-Type` SHOULD NOT use `Content-Disposition: inline`.

Yes, it's a SHOULD NOT rather than a MUST NOT, but if our one and only attempt to follow that SHOULD failed, then it seems pretty valueless.

Author: I feel this thread is going in circles, or at least contributing noise after a conclusion has been agreed to. Namely, it's clear that everyone (except for maybe me) wants the MUST dropped to a SHOULD, which can happen easily.

What remains is whether this is blocking for the FCP to continue collecting checkboxes. Given the vibrancy of debate here, it sounds *highly* blocking.

@richvdh : please raise this as a concern so it doesn't get lost.

I'm interested in hearing opposing viewpoints to that conclusion at this point.

Reviewer 6: > > To effect this layered protection, servers which do not sniff for `Content-Type` SHOULD NOT use `Content-Disposition: inline`.
> 
> Yes, it's a SHOULD NOT rather than a MUST NOT, but if our one and only attempt to follow that SHOULD failed, then it seems pretty valueless.

Ah, I failed to connect that an additional clause that was needed. I agree with @davidegirardi assessment above though. I would drop this clause as a requirement.

`Content-Type` additionally becomes a required header on both `/download` and `/thumbnail`, as `Content-Disposition`
without `Content-Type` is effectively useless in HTTP.

For clarity, a server is *not* required to use `inline` on `/download`. Clients SHOULD assume that a server will
always use `attachment` instead.

## Potential issues

This proposal does not require the usage of `inline` on `/download`, making it harder for IRC and similar
bridges to rely on "pastebin" behaviour. For example, when a large message is posted on the Matrix side of
the bridge, the IRC bridge might upload it as a text file due to limits on the IRC side. Ideally, that text

> file would be rendered inline by the serer. Bridges are encouraged to use "proxy APIs" to serve the text

Reviewer 4: ```suggestion
file would be rendered inline by the server. Bridges are encouraged to use "proxy APIs" to serve the text
```

file instead, where they can better control the user experience.

## Alternatives

No major alternatives identified.

## Security considerations

> This MSC fixes theoretical security issues relating to Cross-Site Scripting and similar. No new security issues

Author: Arguably none. If a server decided to show `application/x-evil` with an `inline` disposition, the browser might be vulnerable to executing that. By forcing `attachment` though, the browser never 'runs' the content and just dumps it.

If a server's CSP and general guidance around not having your web client at the same domain as your media repo are honoured, clients should never run into issues. 

Will spell this out more.

# MSC2702: `Content-Disposition` usage in the media repo

Clients and servers often disagree on how the `Content-Disposition` header should work with respect
to uploads. This proposal aims to settle the disagreement while standardizing the behaviour.

## Proposal

The `Content-Disposition` header becomes required on `/download` and `/thumbnail`, with `/download`. When
a `filename` was supplied to `/upload`, `/download`'s `Content-Disposition` MUST have a filename attribute.

> For `/thumbnail`, the `Content-Disposition` MUST be `inline` (with optional server-generated filename).

Author: fwiw, rationale is largely semantics. It's uncommon that the thumbnail will appear outside of an `img` tag, but it does happen thanks to some clients. Therefore, we just make it `inline` and call it good.

For `/download`, the attribute MUST be one of `inline` or `attachment`. The server SHOULD intelligently
decide which of the two to use, such as by implying all `image/*` content types are `inline` and binary
blobs are `attachment`s.

To allow for clients to specifically request a behaviour, a new query string parameter is added to
`/download` to let clients pick explicitly which one they would like: `asAttachment`. When not
supplied, the server's default processing is to be used. When supplied as `true`, the server MUST
use `Content-Disposition: attachment` (with possible filename). When supplied as `false`, the server
MUST use `Content-Disposition: inline` (with possible filename).

> This proposal does not advocate to let clients specify the entire header out of caution.

Reviewer 8: Looking at [the MDN article](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition), I don't think clients can even specify desired content disposition as an HTTP header for this purpose, unless there's some other header?

Author: this is more of a nod to not allowing `?contentDisposition=plz no` which would get translated verbatim to `Content-Disposition: plz no`

## Potential issues

Servers could "intelligently" decide to always use `inline` by default (as they currently do). This
is acceptable under this proposal, though not advocated by the author.

## Alternatives

We could settle the debate by picking one or the other, however from past discussions not everyone
would be happy with the result. Currently clients link to the `/download` endpoint as a literal "download
this" button (implying they'd like `attachment`) whereas others use the endpoint as a fallback, allowing
users to click on it to view the content in their browser. We could introduce an additional `/preview`
or `/view` endpoint which changes the implicit behaviour of the header, however this solution feels like
unneeded endpoint sprawl.

## Security considerations

None known.

## Unstable prefix

> While this MSC is not included in a spec release, implementations should use `org.matrix.msc2702` as a

Reviewer 5: ```suggestion
Before this MSC is merged, implementations should use `org.matrix.msc2702` as a
```

:)

Author: *the legacy shall live on*

```suggestion
While this MSC is not considered stable, implementations should use `org.matrix.msc2702` as a
```

## Security considerations

> This MSC fixes a theoretical Cross-Site Scripting issue where a browser improperly handling Content
> Security Policy headers or sandboxes *may* reveal sensitive information about the user, if the user's
> client is hosted at the same domain as the media download URL. This is mitigated by only suggesting
> `inline` as a disposition on content types which are not likely to execute code within the browser
> upon being viewed. A browser may still have further bugs which reveal information, though those issues
> quickly become impractical for the Matrix specification to mitigate.

Reviewer 2: I think this sounds too far fetched while it's in fact more practical. I'd rephrase to:
```suggestion
This MSC fixes a possible Cross-Site Scripting issue in enviroments with insecure configurations.
For example, if a Matrix web client is also hosted on the same domain as the media download URL.
```

Author: The mitigation explanation seems to have disappeared in the suggestion :(

Reviewer 2: I don't follow, what's missing exactly?

Author: "This is mitigated..." and onwards - I'll apply the diff manually
